design-docs/incremental-build.md

```diff
@@ -243,3 +243,109 @@ When the `TaskHistory` gets persisted, it adds the current task execution to the
+
+**Example:**
+
+A user compiles two Java source files, A and B, with Gradle version X via `compileJava`. The produced output are the class files C and D. Those class files might be packaged into
+ a JAR file via the `jar` task.
+
+Let's assume the user upgrades the version of Gradle to Y _and_ one of the input files is moved or deleted e.g. B is deleted (which might happen upon a branch change or refactoring of the code). 
+Gradle does not have a task history for the task `compileJava` as `.gradle` directory is based on the used Gradle version e.g. `./gradle/3.1` vs. `.gradle/3.2`. As a result `compileJava` would compile
+source file A but not B as it doesn't exist anymore. The output directory of the compilation task would still contain A and B which leads to an incorrect result.
+
+The goal of this story is to 1) detect this situation and 2) clean up any stale output.
+
+### User visible changes
+
+Gradle detects and removes stale Java class files after a Gradle version change. Further processing of output files (e.g. by the `jar` task) does not result in faulty behavior.
+
+### Implementation
+
+- Establish a registry implementation that allows for registering strategies with the purpose of cleaning up outputs produced by previous builds.
+    - The registry is not specific to tasks. It's rather a global concept that allows for registering clean up strategies for any "thing" in Gradle.
+    - The registry is going to be an internal concept and is not going to be exposed via a public API.
+    - The registry needs to be instantiated by Gradle's global service registry. 
+    - When a build is executed, the registry is contacted and provides all registered strategies.
+    - If a strategy indicates that clean up is needed then delete all registered outputs.
+
+<!-- -->
+
+    public interface BuildOutputCleanupStrategy {
+        /**
+         * Determines if cleanup is needed based on provided Gradle instance.
+         */
+        boolean needsCleanup(Gradle gradle);
+    }
+    
+    public interface BuildOutputCleanup {
+        /**
+         * Returns the clean up strategy.
+         */
+        BuildOutputCleanupStrategy getStrategy();
+        
+        /**
+         * Returns the outputs that need to be cleaned if needed.
+         */
+        List<File> getOutputs();
+    }
+
+    public interface BuildOutputCleanupRegistry {
+        /**
+         * Registers clean up strategy for given outputs.
+         */
+        void registerCleanup(BuildOutputCleanup cleanups);
+        
+        /**
+         * Returns all clean ups.
+         */
+        List<BuildOutputCleanup> getCleanups();
+    }
+
+- Provide an implementation for a clean up strategy that identifies if task history is available.
+    - Create an instance of the implementation in Gradle's global service registry.
+    - Check if the `.gradle/<version>` directory exists for current version of Gradle used for build.
```

Comment:
This check only deals with the _first_ use of a given version. This can be a problem when I'm, for example, switching between branches with different set of source files and different Gradle versions.

This might be fine as a place to start, but we won't be able to use this to clean up buildSrc. Another option would instead be to reuse the buildSrc approach for all builds. Here, we basically persist the Gradle version to a cross-version file and trigger a clean up when this version changes (or there is no version).